synapse/module_api/__init__.py

```diff
@@ -1386,3 +1396,65 @@ async def remove_room_from_public_room_list(self, room_id: str) -> None:
             room_id: The ID of the room.
         """
         await self._store.set_room_is_public(room_id, False)
+
+
+class AccountDataManager:
+    """
+    Allows modules to manage account data.
+    """
+
+    def __init__(self, hs: "HomeServer") -> None:
+        self._hs = hs
+        self._store = hs.get_datastores().main
+        self._handler = hs.get_account_data_handler()
+
+    def _validate_user_id(self, user_id: str) -> None:
+        """
+        Validates a user ID is valid and local.
+        Private method to be used in other account data methods.
+        """
+        user = UserID.from_string(user_id)
+        if not self._hs.is_mine(user):
+            raise ValueError(
+                f"{user_id} is not local to this homeserver; can't access account data for remote users."
+            )
+
+    async def get_global(self, user_id: str, data_type: str) -> Optional[JsonDict]:
+        """
+        Gets some global account data, of a specified type, for the specified user.
+
+        The provided user ID must be a valid user ID of a local user.
+
+        Added in Synapse v1.57.0.
+        """
+        self._validate_user_id(user_id)
+
+        data = await self._store.get_global_account_data_by_type_for_user(
+            user_id, data_type
+        )
+        # We clone to prevent the module accidentally mutating the dict that
+        # lives in the cache, as that could introduce nasty bugs.
+        return copy.deepcopy(data)
```

Comment:
I was half-and-half on this one. I obviously don't want modules to mutate internal data.
But if you're going to give the module a clone anyway, what does freezing it buy you? They already can't damage anything, since it's a clone.

Maybe it makes more sense anyway. I guess I'll do it for now, it'll be easier to revert later than the other way around if we're not keen.

Reply:
Freezing it makes it immutable, so will raise errors instead of allowing mutation but doing nothing. That seems clearer to me?